File: csharp-project/RekindleBookStoreAqa/RekindleTaf/Tests/Api/BookstoreProductServiceTest.cs
using System.Collections.Generic;
using System.Net;
using FluentAssertions;
using NUnit.Extension.DependencyInjection;
using RekindleTaf.Controllers;
using RekindleTaf.Exceptions;
using RekindleTaf.Model;
using RekindleTaf.Model.Factory;
using RekindleTaf.Utils;

namespace RekindleTaf.Tests.Api;

[DependencyInjectingTestFixture]
public class BookstoreProductServiceTest(
    IFactoryKit factoryKit,
    IBookstoreController bookstoreController,
    IBookstoreProductController productController) : Base
{
    private readonly List<Product> Products = new();
    private Bookstore bookstore;

    [OneTimeSetUp]
    public void Setup()
    {
        bookstore = factoryKit.Factory<Bookstore>(FactoryKey.Bookstore).Create(BookstoreKey.StdBookstoreActive);
        bookstore = bookstoreController.PostSuccessfully(bookstore);
    }
    
    [OneTimeTearDown]
    public void Cleanup()
    {
        Products.ForEach(productController.DeleteSuccessfully);
        bookstoreController.DeleteSuccessfully(bookstore);
    }
    
    [Test]
    public void Create_NewProduct_ShouldReturnOk()
    {
        var product = factoryKit.Factory<Product>(FactoryKey.Product).Create(ProductKey.StdProductAvailable);
        product.BookstoreId = bookstore.Id;
        product = productController.PostSuccessfully(product);
        Products.Add(product);
        product.Id.Should().NotBeEmpty();
    }
    
    [Test]
    public void Get_NewProduct_ShouldReturnOk()
    {
        var product = factoryKit.Factory<Product>(FactoryKey.Product).Create(ProductKey.StdProductAvailable);
        product.BookstoreId = bookstore.Id;
        product = productController.PostSuccessfully(product);
        Products.Add(product);
        var fetchedProduct = productController.GetSuccessfully(product.Id!);
        fetchedProduct.Id.Should().Be(product.Id);
        fetchedProduct.Name.Should().Be(product.Name);
        fetchedProduct.Available.Should().Be(product.Available);
        fetchedProduct.Price.Should().Be(product.Price);
    }
    
    [Test]
    public void Get_AllProduct_ShouldReturnOk()
    {
        var products = productController.GetAllSuccessfully();
        products.Should().NotBeEmpty();
    }
    
    [Test]
    public void Update_NewProduct_ShouldReturnNoContent()
    {
        var product = factoryKit.Factory<Product>(FactoryKey.Product).Create(ProductKey.StdProductAvailable);
        product.BookstoreId = bookstore.Id;
        product = productController.PostSuccessfully(product);
        Products.Add(product);
        decimal newPrice = new decimal(999.99);
        product.Price = newPrice;
        productController.PutSuccessfully(product);
        var fetchedProduct = productController.GetSuccessfully(product.Id!);
        fetchedProduct.Price.Should().Be(product.Price);
    }
    
    [Test]
    public void Delete_NewProduct_ShouldReturnNoContent()
    {
        var product = factoryKit.Factory<Product>(FactoryKey.Product).Create(ProductKey.StdProductAvailable);
        product.BookstoreId = bookstore.Id;
        product = productController.PostSuccessfully(product);
        productController.DeleteSuccessfully(product);
        var response = productController.Get(product.Id!).Result;
        response.StatusCode.Should().Be(HttpStatusCode.NotFound);
        response.GetNode("errorMessage").Should().Be(ErrorMessage.ProductNotFound);
    }
}

File: typescript-project/rekindle-book-store-aqa/tests/bookstore-product-service.test.ts
import {afterAll, beforeAll, describe, expect, test} from '@jest/globals';
import {Bookstore, Product} from '@/model/domain-model';
import {Factory, FactoryKey, FactoryKit} from '@/model/factory';
import {container} from 'tsyringe';
import {HttpStatusCode} from 'axios';
import {BookstoreKey} from "@/model/bookstore-factory";
import {ProductKey} from "@/model/product-factory";
import {ErrorMessage, randomPriceLessThanThousand} from "../src/utils/utils";
import {BookstoreCrudController, BookstoreProductCrudController} from "@/rest/controllers/rest-controllers";
import {RestClient} from "@/rest/rekindle-rest-client";

describe('Bookstore Product Service Test', () => {
    let factoryKit: FactoryKit;
    let bookstoreController: BookstoreCrudController;
    let productController: BookstoreProductCrudController;
    let productFactory: Factory<Product>;
    let bookstore: Bookstore;
    const products: Product[] = [];

    beforeAll(async () => {
        await container.resolve<RestClient>('RestClient').setupToken();
        factoryKit = container.resolve('FactoryKit');
        productFactory = factoryKit.factory<Product>(FactoryKey.product);
        bookstoreController = container.resolve('BookstoreCrudController');
        productController = container.resolve('BookstoreProductCrudController');

        const bookstoreFactory: Factory<Bookstore> = factoryKit.factory<Bookstore>(FactoryKey.bookstore);
        bookstore = bookstoreFactory.create(BookstoreKey.stdBookstoreActive);
        bookstore = await bookstoreController.postSuccessfully(bookstore);
    });

    afterAll(async () => {
        if (products.length === 0) return;
        for (const product of products) {
            bookstoreController.deleteSuccessfully(product);
        }
    });

    test('Create New Product Should Return OK', async () => {
        const product: Product = productFactory.create(ProductKey.stdProductActive);
        product.bookstoreId = bookstore.id;

        const createdProduct = await productController.postSuccessfully(product);
        products.push(createdProduct);

        expect(createdProduct.id).toBeDefined();
    });

    test('Get New Product Should Return OK', async () => {
        const product: Product = productFactory.create(ProductKey.stdProductActive);
        product.bookstoreId = bookstore.id;

        const createdProduct = await productController.postSuccessfully(product);
        products.push(createdProduct);

        const fetchedProduct = await productController.getSuccessfully(createdProduct.id);
        expect(fetchedProduct).toMatchObject({
            id: createdProduct.id,
            name: createdProduct.name,
            available: createdProduct.available,
            price: createdProduct.price
        });
    });

    test('Get All Products Should Return OK', async () => {
        const allProducts = await productController.getAllSuccessfully();
        expect(allProducts).not.toHaveLength(0);
    });

    test('Update Product Should Return No Content', async () => {
        const product: Product = productFactory.create(ProductKey.stdProductActive);
        product.bookstoreId = bookstore.id;

        const createdProduct = await productController.postSuccessfully(product);
        products.push(createdProduct);

        const updatedPrice = randomPriceLessThanThousand();
        createdProduct.price = updatedPrice;
        await productController.putSuccessfully(createdProduct);

        const updatedProduct = await productController.getSuccessfully(createdProduct.id);
        expect(updatedProduct.price).toBe(updatedPrice);
    });

    test('Delete Product Should Return No Content', async () => {
        const productFactory: Factory<Product> = factoryKit.factory<Product>(FactoryKey.product);
        const product: Product = productFactory.create(ProductKey.stdProductActive);
        product.bookstoreId = bookstore.id;

        const createdProduct = await productController.postSuccessfully(product);

        await productController.deleteSuccessfully(createdProduct);

        await productController.get(createdProduct.id).catch((error) => {
            expect(error.response?.status).toBe(HttpStatusCode.NotFound);
            expect(error.response?.data?.errorMessage).toBe(ErrorMessage.productNotFound);
        });

    });
});
